File: frontend/src/app/models/tasks/monitoring-task-configuration-request.ts
import { TaskType } from './task-type';

export interface PingTaskConfigurationRequest {
  type: TaskType.PING;
  timeoutMs: number;
}

export interface TelnetTaskConfigurationRequest {
  type: TaskType.TELNET;
  port: number;
  timeoutMs: number;
}

export interface HttpHealthcheckTaskConfigurationRequest {
  type: TaskType.HTTP_HEALTHCHECK;
  port: number;
  path: string;
  timeoutMs: number;
}

export type MonitoringTaskConfigurationRequest =
  | PingTaskConfigurationRequest
  | TelnetTaskConfigurationRequest
  | HttpHealthcheckTaskConfigurationRequest;

File: web-api/src/main/java/pl/sgorski/nethelt/webapi/features/monitoring_task/dto/request/MonitoringTaskConfigurationRequest.java
package pl.sgorski.nethelt.webapi.features.monitoring_task.dto.request;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import pl.sgorski.nethelt.webapi.features.monitoring_task.dto.request.configuration.HttpHealthcheckTaskConfigurationRequest;
import pl.sgorski.nethelt.webapi.features.monitoring_task.dto.request.configuration.PingTaskConfigurationRequest;
import pl.sgorski.nethelt.webapi.features.monitoring_task.dto.request.configuration.TelnetTaskConfigurationRequest;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
@JsonSubTypes({
  @JsonSubTypes.Type(value = PingTaskConfigurationRequest.class, name = "PING"),
  @JsonSubTypes.Type(value = TelnetTaskConfigurationRequest.class, name = "TELNET"),
  @JsonSubTypes.Type(
      value = HttpHealthcheckTaskConfigurationRequest.class,
      name = "HTTP_HEALTHCHECK")
})
public interface MonitoringTaskConfigurationRequest {}
